Types Team Update and Roadmap (#1346)
rust-lang/blog.rust-lang.org: posts/2024-06-20-types-team-update.md

## Organizational updates

First, a huge welcome to the two new members to team since the announcement post: [@BoxyUwU] and [@aliemjay]. [@BoxyUwU] has been doing a lot of work on const generics and made significant contributions to the design of the next generation trait solver. [@aliemjay] has been working on some very subtle improvements to opaque types - `impl Trait` - and to borrow checking. They are both invaluable additions to the team.

We also organized another in-person Types Team meetup last October, immediately prior to EuroRust. We discussed the state of the team, looked at current implementation challenges and in-progress work, and reviewed and updated [the roadmap from the previous meetup][PrevRoadmap]. Most of this will be covered in this blog post.

Finally, as discussed in the [RFC](https://rust-lang.github.io/rfcs/3254-types-team.html), we would like to have leads rotate out regularly, largely to help share the burden and experience of leads' work. So with that being said, [@nikomatsakis](https://github.com/nikomatsakis) is rotating out and [@lcnr](https://github.com/lcnr) is joining to co-lead alongside [@jackh726](https://github.com/jackh726/).

> ## Roadmap progress and major milestones

Reviewer: Another major accomplishment that could be worth mentioning is:

- https://github.com/rust-lang/rust/pull/116733

That seems on-par in terms of significance with some of the other things mentioned.

Author: while this is a very nice and necessary change, it does feel slightly "smaller in scope" to me than the other opaque type changes mentioned here. tbh I also just don't know how I would best add that to the existing document.

### The next-generation trait solver

There has been [a lot of work][NewSolver] on the [next-generation trait solver][InitiativeRepo].
The initiative posted [a separate update][InitiativeUpdate] at the end of last year. While
we would have liked to [stabilize its use in coherence][StabilizeNS] a few months ago,
this surfaced additional small behavior regressions and hangs, causing delays. We are working on fixing these issues and intend to merge the stabilization PR soon. We are getting close to compiling the standard library
and the compiler with the new solver enabled everywhere, after which will be able to run
crater to figure out the remaining issues. We expect there to be a long tail of minor issues
and behaviorial differences from the existing implementation, so there's still a lot to do
here. There are also open design questions which we will have to resolve before stabilizing
the new implementation.

### Async and `impl Trait`

We stabilized `async`-fn in traits (AFIT) and return-position `impl Trait` in

> traits (RPTIIT) in version 1.79 thanks to a significant effort by [@compiler-errors] and

Reviewer: ```suggestion
traits (RPTIIT) in version 1.75 thanks to a significant effort by [@compiler-errors] and
```

AFIT and RPITIT stabilized in Rust 1.75.

[@spastorino]. [@cjgillot] greatly improved the way generators, and therefore async functions,
are represented in the type system[^107421]. This allowed us to support recursive
`async`-functions without too much additional work[^117703].

Designing the next-generation trait solver surfaced issues and future-compatability challenges
of our type-alias `impl Trait` (TAIT) implementation using the old trait solver. We are
currently reworking the design and implementation. [@oli-obk] is spear-heading this effort.
This also impacts RPIT edge-cases, forcing us to be careful to avoid accidental breakage.

> There have also been some new language design questions for TAIT, so we will likely first

Reviewer: ```suggestion
There are some open language design questions for TAIT, so we will likely first
```

The questions are open but not really new.

> There have also been some new language design questions for TAIT, so we will likely first
> stabilize associated type position `impl Trait` (ATPIT) as it avoids these language design

Reviewer: ```suggestion
There have also been some new language design questions for TAIT, so we plan to first
stabilize associated type position `impl Trait` (ATPIT) as it avoids these language design
```

It's rather more than likely at this point, so perhaps let's just say that's the plan.

> stabilize associated type position `impl Trait` (ATPIT) as it avoids these language design
> issues while still providing the same capabilities.

Reviewer: ```suggestion
stabilize associated type position `impl Trait` (ATPIT) as it avoids these
questions while still closing the expressiveness gap.
```

I'd suggest weakening the language from "issues" to "questions" (as was done earlier in the sentence), removing some repeated verbiage, and saying that ATPIT closes the expressiveness gap rather than offering the same "capabilities".

(Maybe it does offer the same capabilities, but it's more obviously true that it closes the expressiveness gap.)

everywhere by default and to fully replace the existing implementation. We are currently
finalizing the stabilization of [its use in coherence checking][StabilizeNS]. This should
already fix multiple unsound issues and fix a lot of smaller issues and inconsistencies of
the current implementation. See the stabilization report for more details.

We are also working on extracting its implementation into a separate library
outside of the compiler itself. We would like to share the trait solver with
rust-analyzer by the end of this year. They currently use [chalk] which is no longer
actively maintained. Using the next-generation trait solver in rust-analyzer
should result in a lot of additional testing for the solver and while also improving
the IDE experience by positively impacting performance and correctness.

We intend to slowly roll out the solver in other areas of the compiler until we're able
to fully remove the existing implementation by the end of 2025. This switch will fix
multiple unsound issues by itself and will unblock a significant amount of future work.
It will generally cleanup many rough edges of the type system, such as associated types

> in binders. There are many unsound issues which can only be fixed once we exclusively

Reviewer 2: I would expect most people reading this not to know what a binder is

Using it to model parts of the type system has already been incredibly impactful and we want
to build on that. We are working on more effective testing of `a-mir-formality` by enabling its
use for actual Rust code snippets and by adding fuzzing support. This will allow us to gain
additional confidence in our model of the type system and will guide its future development.

We plan to fully formalize some components of the type system this year. Coherence is fairly
self-contained, very subtle, and soundness-critical. This has prevented us from making significant
improvents to it in the past. We also intend to formalize coinductive trait semantics, which are
difficult to reason about and necessary to fix many longstanding soundness issues.

### Language changes and polonius

* [Associated Type Position Impl Trait (ATPIT) project goal](https://rust-lang.github.io/rust-project-goals/2024h2/ATPIT.html)
* [Polonius on Nightly project goal](https://rust-lang.github.io/rust-project-goals/2024h2/Polonius.html)

We intend to get the internal implementation of opaque types ready for the stabilization

> of TAIT and APTIT this year. We are also hoping to land a significant improvements our

Reviewer 3: Did you mean ATPIT?

Author: yes

> of TAIT and ATPIT this year. We are also hoping to land a significant improvements our
> handling of associated types in coherence checking this year.

Reviewer: ```suggestion
of TAIT and ATPIT this year. We are also hoping to land significant improvements to our
handling of associated types in coherence checking this year.
```

Our other goal is to get [Polonius], the next generation borrow checker, available on nightly, which would put us in a position to stabilize in 2025 once we have time to do more optimization and testing.

[polonius]: https://blog.rust-lang.org/inside-rust/2023/10/06/polonius-update.html

We also intend to support the development of other language features, such as `async`-closures, which are part of the [async project goal](https://rust-lang.github.io/rust-project-goals/2024h2/async.html),
and `dyn`-trait upcasting, which will hopefully get stabilized in the near future.

## Roadmap

### EOY 2024

- next-generation trait solver
    - stable in coherence
    - used by rust-analyzer

> - TAIT/ATPIT stabilized

Reviewer: ```suggestion
- ATPIT stabilized
```

(Speaking as the language-side owner of this here:) It would surprise me (pleasantly, but still) if we stabilized TAIT in 2024.  It'd probably be better to not overpromise here and to just say that we're planning to stabilize ATPIT in 2024, which we are.
